### PR TITLE
i#4083 func trace: Add func_view tool

### LINKED ISSUE #4083: Add drcachesim function trace viewing tool
The proposal is to create a simple analysis tool which will provide both a sequential view of the function trace data recorded in a drcachesim trace (through -record_heap or -record_function) as well as aggregate counts.

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new func_view tool for visualizing and analyzing function calls
recorded in a drcachesim trace.  It does two things in its initial
form: it reports the summary count of calls and returns for every
function traced, and if a new option op_show_func_trace is set (on by
default), it prints a linear sequence of every call and return with
return addresses, argument values, and return values.  Each is
indented to enable a clear picture of the call sequence.

Adds a new option op_funclist_file to specify the file's path, though
normally it will be auto-found.

Adds raw2trace_directory_t support for parsing the funclist file.

Adds example output for the new tool to the docs.  Adds a brief
section to the docs on function tracing in general, which was
previously missing.

Adds a func_view test which runs fib(5) and ensures we get the nested
function trace correct.

Fixes #4083

### CODE REVIEW

File: clients/drcachesim/tools/func_view.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    bool knob_full_trace_;
+    unsigned int knob_verbose_;
+
+    std::unordered_map<int, std::set<std::string>> id2name_;
+
+    std::string funclist_file_path_;
+
+    // The keys here are shard index for parallel, tid for serial.
+    std::unordered_map<memref_tid_t, shard_data_t *> shard_map_;
+    // This mutex is only needed in parallel_shard_init.  In all other accesses to
```
Comment (s-kanev) [inline clients/drcachesim/tools/func_view.h]: Your call, might still be worth grabbing in the other cases so it's easier to make tsan happy in the future.

File: clients/drcachesim/tools/func_view.cpp [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+#include "dr_api.h"
+#include "func_view.h"
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+#include <assert.h>
+
+#define EXPANDSTR(x) #x
+#define STRINGIFY(x) EXPANDSTR(x)
```
Comment (s-kanev) [inline clients/drcachesim/tools/func_view.cpp]: unused

File: clients/drcachesim/tools/func_view.cpp [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+std::string
+func_view_t::initialize()
+{
+    std::vector<std::pair<int, std::string>> entries;
+    raw2trace_directory_t directory_;
+    std::string error =
+        directory_.initialize_funclist_file(funclist_file_path_, &entries);
+    if (!error.empty())
+        return "Failed to read " + funclist_file_path_ + error;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/func_view.cpp]: Done.

File: clients/drcachesim/tools/func_view.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    return !knob_full_trace_;
+}
+
+void *
+func_view_t::parallel_shard_init(int shard_index, void *worker_data)
+{
+    auto shard_data = new shard_data_t;
+    std::lock_guard<std::mutex> guard(shard_map_mutex_);
+    shard_map_[shard_index] = shard_data;
+    return reinterpret_cast<void *>(shard_data);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/func_view.cpp]: Dropped.

File: clients/drcachesim/tools/func_view.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tools/func_view.cpp]: Thanks, relic from earlier code.  Removed.

File: clients/drcachesim/tools/func_view.h (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tools/func_view.h]: This is how the other tools use the lock as well and tsan does not complain.  I just ran this new one (internally w/ adapters) under tsan as well and it does not report any problems. (Confirmed that removing the lock from parallel_shared_init does result in data races reported.)  I am inclined to leave it to match the others and avoid extra locks.

**Comment by derekbruening** [comment]
Well I got the test to pass on my local Windows machine, but on Appveyor it is hitting a mysterious crash with no further information:

```
[00:24:03] 205: Running postcmd |C:/projects/dynamorio/build/build_debug-internal-64/clients/bin64/drcachesim.exe;-indir;C:/projects/dynamorio/build/build_debug-internal-64/suite/tests/tool.drcacheoff.func_view.common.fib.exe.07132.7209.dir;-simulator_type;func_view|
[00:24:03] 205: CMake Error at C:/projects/dynamorio/suite/tests/runmulti.cmake:106 (message):
[00:24:03] 205:   *** postcmd failed (Access violation): ***
[00:24:03] 205: 
[00:24:03] 205: Call Stack (most recent call first):
[00:24:03] 205:   C:/projects/dynamorio/suite/tests/runmulti.cmake:118 (process_cmdline)
[00:24:03] 205: 
[00:24:03] 205: 
[00:24:03] 205/239 Test #205: code_api|tool.drcacheoff.func_view ...........................***Failed   14.80 sec
```
A crash is disturbing.  But If I can't reproduce I will probably have to disable the test on Windows and put this PR in and consider the crash separately.

**Comment by derekbruening** [comment]
Hmm, Mac is starting to have strange results.  Twice now the log is truncated, but there's no timeout message, and the time is not some even number.

**Comment by derekbruening** [comment]
A64 is #3742 threadexit2

**Pull request merged**